### ISSUE TITLE
does this only work with homey?

### ISSUE THREAD

**Issue by mark-hahn** [AdyRock/SwitchBotBLEHub#4]
I saw "Welcome to the ESP32 SwitchBot BLE hub for Homey" in the code.   I just wanted to use it as a rest api for my custom app.   Can I just ignore the homey stuff?

**Comment by mark-hahn** [comment]
After studying it more it appears that this offers an api identical to the switchbot cloud api.  Is this correct?  If so then this is awesome.  And I can ignore the homey stuff, right?

I already access the cloud with my code.  I got killed by the 10,000 hits/day.  Now I can just replace the base url and hit the esp32 as fast as I want.

**Comment by AdyRock** [comment]
The idea is that the ESP32 pushes notifications to the host ( normally Homey). My Homey app registers the callback when it detects the ESP32 via a broadcast and, at the same time, requests the current state of the devices.
So, you could poll the ESP32 for data, but it is more efficient to use the callback as then there is only traffic when something changes. Also, the data format is different from the cloud API, as it is more like the BLE data. My Homey app supports both cloud and BLE via Homey's built-in BLE, so the idea is to extend the BLE range of Homey by getting BLE compatible information via WiFi.
If I find time, I will document the API, but it was designed as a plug and play device for Homey.

**Comment by mark-hahn** [comment]
So the short answer _is_ that it is meant for homey now.  I have no desire to pay for more hardware so homey is not something I'm interested in. 

If I knew your api and adapted the changes to my code then would your existing SwitchBotBLEHub code serve my purpose?  I will study your code and see if I can reverse engineer the API.  If successful I would be happy to add the api docs to your repo.  

Do you know of any other software that could help me directly access switchbot ble?

**Comment by AdyRock** [comment]
It could certainly be used if you adapt your code as there is nothing specific to Homey. 
If your code can run a server to listen for messages, then it will be the most efficient method for both ends.
If not, then you can poll the API, and the only limit will be the processing power of ESP32 to service the calls.

**Comment by mark-hahn** [comment]
That's good to hear. After thinking about it I realized I don't have to know the whole API.  I can just use urls I find in your code and adapt to the results.  

I am going to do polling.  Hitting it every 20 seconds would put no load on my server, the network, or the esp32.  Actually, I only need the one command to poll the device status.  I already know the device IDs and names from the switchbot app.  (Keep it simple). 

Thanks for your help and your software.  Wish me luck.  I'll keep you posted.

**Comment by mark-hahn** [comment]
Things are going great.  The binary didn't load onto my esp32-D but I was able to compile to it and it runs great.  But I'm having a problem with the return data from the /device url.  Here is the url ...
```
http://192.168.1.86/api/v1/device?address=FB3A47261098
```

And here is the return data ...
```
 {
    hubMAC: '64:90:f1:18:2e:c8',
    serviceData: { error: 'Invalid index 255' }
  }
```
It must be finding the device because it has the correct hubMAC.  Do have any idea what is wrong with the serviceData?

**Comment by mark-hahn** [comment]
I just had a thought.  I'm not calling the /devices url.  Is that necessary?

**Comment by mark-hahn** [comment]
When I added the /devices call I discovered that all the data I need is returned from all devices.  So I'm using that and I removed the /device calls.  So now I've got everything working!

Thanks for answering all my questions and for the great software.

**Comment by AdyRock** [comment]
It's great that you have it working. 
The issue you have with the /device call is the device id, as that needs to be the BLE uuid (which is the BLE mac address). The id you passed in is the cloud generate one, which the BLE has no knowledge of.
So basically, the "invalid index..." error means it couldn't find the device.

Edit: Actually, I just looked at your id again, and now I'm not sure, so maybe it's just the format.

**Comment by mark-hahn** [comment]
thanks

**Comment by AdyRock** [comment]
I checked, and the id lookup uses a case insensitive string comparison, so the id would need to be in the colon notation.

**Comment by mark-hahn** [comment]
For those who stumble into this thread I have an FYI to add.

When I compiled the source for use in ae esp32-D I had a problem.  Every day or so the hub would quit responding for 30 mins or so.  I was using only the /devices calls with polling every 20 secs.  I suspected that the polling was somehow screwing things up.  So I rewrote my app to use callbacks.  Unfortunately this didn't help.

The readme said that this software was tested on an esp32-S.  So I bought one of those and used the pre-compiled binary.  This has been working fine for 3 days so far.
